File: sdks/node/src/plugins/augment-contract.ts
import os from 'node:os';
import path from 'node:path';

export const AGENT_KEY = 'augment';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_MILLISECONDS = 10_000;

const WRAPPER_EXTENSION = process.platform === 'win32' ? '.cmd' : '.sh';
export const GUARD_WRAPPER = `augment-guard${WRAPPER_EXTENSION}`;
export const AUDIT_WRAPPER = `augment-hook${WRAPPER_EXTENSION}`;

const TOOL_EVENTS = new Set(['PreToolUse', 'PostToolUse']);
const SESSION_EVENTS = new Set([
  'Stop',
  'SessionStart',
  'SessionEnd',
  'Notification',
  'PromptSubmit',
]);

export type JsonObject = Record<string, unknown>;

export interface AugmentHandler extends JsonObject {
  readonly type: 'command';
  readonly command: string;
  readonly args?: string[];
  readonly timeout?: number;
}

export interface AugmentGroup extends JsonObject {
  readonly matcher?: string;
  readonly hooks: AugmentHandler[];
}

export type AugmentHooks = Record<string, AugmentGroup[]>;

export interface AugmentDocument {
  readonly exists: boolean;
  readonly configPath: string;
  readonly root: JsonObject;
  readonly hooks: AugmentHooks;
}

export interface RuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
  readonly guardWrapperPath: string;
  readonly auditWrapperPath: string;
}

interface ParsedArgument {
  readonly value: string;
  readonly next: number;
}

export function isObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasOwn(value: JsonObject, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, key);
}

function quotePosix(value: string): string {
  return `'${value.replaceAll("'", `'"'"'`)}'`;
}

function quoteWindows(value: string): string {
  return `"${value.replaceAll('"', '\\"')}"`;
}

function quoteBatch(value: string): string {
  return `"${value.replaceAll('%', '%%')}"`;
}

function readWindowsArgument(command: string, start: number): ParsedArgument | undefined {
  if (command[start] !== '"') return undefined;
  let value = '';
  for (let index = start + 1; index < command.length; index += 1) {
    if (command[index] === '\\' && command[index + 1] === '"') {
      value += '"';
      index += 1;
      continue;
    }
    if (command[index] === '"') return { value, next: index + 1 };
    value += command[index];
  }
  return undefined;
}

const POSIX_APOSTROPHE = `'"'"'`;

function readPosixArgument(command: string, start: number): ParsedArgument | undefined {
  if (command[start] !== "'") return undefined;
  let value = '';
  for (let index = start + 1; index < command.length;) {
    if (command.startsWith(POSIX_APOSTROPHE, index)) {
      value += "'";
      index += POSIX_APOSTROPHE.length;
      continue;
    }
    if (command[index] === "'") return { value, next: index + 1 };
    value += command[index];
    index += 1;
  }
  return undefined;
}

function parseWrapperCommand(command: string): string | undefined {
  const readArgument = process.platform === 'win32' ? readWindowsArgument : readPosixArgument;
  const parsed = readArgument(command, 0);
  return parsed?.next === command.length && parsed.value ? parsed.value : undefined;
}

function samePath(left: string, right: string): boolean {
  const normalizedLeft = path.resolve(left);
  const normalizedRight = path.resolve(right);
  return process.platform === 'win32'
    ? normalizedLeft.toLowerCase() === normalizedRight.toLowerCase()
    : normalizedLeft === normalizedRight;
}

function sameAgentId(left: string, right: string): boolean {
  return process.platform === 'win32' ? left.toLowerCase() === right.toLowerCase() : left === right;
}

export function wrapperPaths(agentId: string): {
  readonly guardPath: string;
  readonly auditPath: string;
} {
  const agentDirectory = path.join(os.homedir(), '.elydora', agentId);
  return {
    guardPath: path.join(agentDirectory, GUARD_WRAPPER),
    auditPath: path.join(agentDirectory, AUDIT_WRAPPER),
  };
}

export function buildHandler(wrapperPath: string): AugmentHandler {
  const quote = process.platform === 'win32' ? quoteWindows : quotePosix;
  return {
    type: 'command',
    command: quote(wrapperPath),
    timeout: HOOK_TIMEOUT_MILLISECONDS,
  };
}

export function buildWrapper(runtimePath: string): string {
  if (process.platform === 'win32') {
    return `@echo off\r\n${quoteBatch(process.execPath)} ${quoteBatch(runtimePath)}\r\nexit /b %errorlevel%\r\n`;
  }
  return `#!/bin/sh\nexec ${quotePosix(process.execPath)} ${quotePosix(runtimePath)}\n`;
}

function managedAgentId(handler: AugmentHandler, wrapperName: string): string | undefined {
  if (handler.type !== 'command'
    || handler.timeout !== HOOK_TIMEOUT_MILLISECONDS
    || typeof handler.command !== 'string'
    || hasOwn(handler, 'args')) return undefined;
  const wrapperPath = parseWrapperCommand(handler.command);
  if (!wrapperPath) return undefined;
  const actualName = path.basename(wrapperPath);
  const namesMatch = process.platform === 'win32'
    ? actualName.toLowerCase() === wrapperName.toLowerCase()
    : actualName === wrapperName;
  if (!namesMatch) return undefined;
  const agentDirectory = path.dirname(wrapperPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) return undefined;
  const agentId = path.basename(agentDirectory);
  return agentId && agentId !== '.' && agentId !== '..' ? agentId : undefined;
}

function validateHandler(
  value: unknown,
  event: string,
  groupIndex: number,
  handlerIndex: number,
): AugmentHandler {
  const label = `Auggie settings handler hooks.${event}[${groupIndex}].hooks[${handlerIndex}]`;
  if (!isObject(value)) throw new Error(`${label} must be an object`);
  if (value.type !== 'command') throw new Error(`${label} type must be "command"`);
  if (typeof value.command !== 'string' || value.command.length === 0) {
    throw new Error(`${label} requires a non-empty command`);
  }
  if (value.args !== undefined
    && (!Array.isArray(value.args) || !value.args.every((argument) => typeof argument === 'string'))) {
    throw new Error(`${label} args must be an array of strings`);
  }
  if (value.timeout !== undefined
    && (typeof value.timeout !== 'number' || !Number.isFinite(value.timeout) || value.timeout <= 0)) {
    throw new Error(`${label} timeout must be a positive finite number`);
  }
  return value as AugmentHandler;
}

function validateMetadata(value: unknown, label: string): void {
  if (!isObject(value)) throw new Error(`${label} metadata must be an object`);
  for (const key of ['includeConversationData', 'includeMCPMetadata', 'includeUserContext']) {
    if (value[key] !== undefined && typeof value[key] !== 'boolean') {
      throw new Error(`${label} metadata.${key} must be a boolean`);
    }
  }
}

function validateGroup(value: unknown, event: string, groupIndex: number): AugmentGroup {
  const label = `Auggie settings group hooks.${event}[${groupIndex}]`;
  if (!isObject(value)) throw new Error(`${label} must be an object`);
  if (SESSION_EVENTS.has(event) && hasOwn(value, 'matcher')) {
    throw new Error(`${label} matcher is only supported for tool events`);
  }
  if (value.matcher !== undefined) {
    if (typeof value.matcher !== 'string') throw new Error(`${label} matcher must be a string`);
    try {
      new RegExp(value.matcher);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`${label} matcher must be a valid regular expression: ${message}`);
    }
  }
  if (value.metadata !== undefined) validateMetadata(value.metadata, label);
  if (!Array.isArray(value.hooks)) throw new Error(`${label} must contain a hooks array`);
  const handlers = value.hooks.map(
    (handler, handlerIndex) => validateHandler(handler, event, groupIndex, handlerIndex),
  );
  return { ...value, hooks: handlers } as AugmentGroup;
}

export function readHooks(root: JsonObject): AugmentHooks {
  if (root.hooks === undefined) return {};
  if (!isObject(root.hooks)) throw new Error('Auggie settings field "hooks" must be an object');
  const hooks: AugmentHooks = {};
  for (const [event, value] of Object.entries(root.hooks)) {
    if (!TOOL_EVENTS.has(event) && !SESSION_EVENTS.has(event)) {
      throw new Error(`Auggie settings field "hooks.${event}" uses an unsupported event`);
    }
    if (!Array.isArray(value)) {
      throw new Error(`Auggie settings field "hooks.${event}" must be an array`);
    }
    hooks[event] = value.map((group, index) => validateGroup(group, event, index));
  }
  return hooks;
}

function removeManaged(
  groups: AugmentGroup[],
  wrapperName: string,
  agentId?: string,
): { readonly groups: AugmentGroup[]; readonly changed: boolean } {
  let changed = false;
  const result: AugmentGroup[] = [];
  for (const group of groups) {
    let groupChanged = false;
    const handlers = group.hooks.filter((handler) => {
      const managedId = managedAgentId(handler, wrapperName);
      const remove = managedId !== undefined
        && (agentId === undefined || sameAgentId(managedId, agentId));
      if (remove) {
        changed = true;
        groupChanged = true;
      }
      return !remove;
    });
    if (!groupChanged) result.push(group);
    else if (handlers.length > 0) result.push({ ...group, hooks: handlers });
  }
  return { groups: result, changed };
}

export function removeManagedHooks(hooks: AugmentHooks, agentId?: string): {
  readonly hooks: AugmentHooks;
  readonly changed: boolean;
} {
  const next = { ...hooks };
  let changed = false;
  for (const [event, wrapperName] of [
    ['PreToolUse', GUARD_WRAPPER],
    ['PostToolUse', AUDIT_WRAPPER],
  ] as const) {
    const result = removeManaged(next[event] ?? [], wrapperName, agentId);
    if (!result.changed) continue;
    changed = true;
    if (result.groups.length > 0) next[event] = result.groups;
    else delete next[event];
  }
  return { hooks: next, changed };
}

function managedIds(groups: AugmentGroup[], wrapperName: string): Set<string> {
  const ids = new Set<string>();
  for (const group of groups) {
    for (const handler of group.hooks) {
      const agentId = managedAgentId(handler, wrapperName);
      if (agentId) ids.add(agentId);
    }
  }
  return ids;
}

export function runtimeContracts(hooks: AugmentHooks): RuntimeContract[] {
  const guards = managedIds(hooks.PreToolUse ?? [], GUARD_WRAPPER);
  const audits = managedIds(hooks.PostToolUse ?? [], AUDIT_WRAPPER);
  const root = path.join(os.homedir(), '.elydora');
  return [...guards]
    .filter((agentId) => [...audits].some((auditId) => sameAgentId(auditId, agentId)))
    .map((agentId) => {
      const agentDirectory = path.join(root, agentId);
      return {
        agentId,
        guardPath: path.join(agentDirectory, GUARD_SCRIPT),
        auditPath: path.join(agentDirectory, AUDIT_SCRIPT),
        guardWrapperPath: path.join(agentDirectory, GUARD_WRAPPER),
        auditWrapperPath: path.join(agentDirectory, AUDIT_WRAPPER),
      };
    });
}

File: sdks/go/cmd/elydora/plugins/augment_contract.go
package plugins

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"path/filepath"
	"reflect"
	"runtime"
	"strings"
)

const (
	augmentAgentKey        = "augment"
	augmentGuardScript     = "guard.js"
	augmentAuditScript     = "hook.js"
	augmentHookTimeout     = float64(10_000)
	augmentPOSIXApostrophe = `'"'"'`
)

var (
	augmentToolEvents = map[string]struct{}{
		"PreToolUse":  {},
		"PostToolUse": {},
	}
	augmentSessionEvents = map[string]struct{}{
		"Stop":         {},
		"SessionStart": {},
		"SessionEnd":   {},
		"Notification": {},
		"PromptSubmit": {},
	}
)

type augmentGroup struct {
	object   map[string]any
	handlers []map[string]any
}

type augmentHooks map[string][]augmentGroup

type augmentDocument struct {
	exists     bool
	configPath string
	root       map[string]any
	hooks      augmentHooks
	raw        []byte
}

type augmentRenderedDocument struct {
	document *augmentDocument
	changed  bool
	next     []byte
	remove   bool
}

type augmentWrapperPaths struct {
	guard string
	audit string
}

type augmentRuntimeContract struct {
	agentID      string
	guardPath    string
	auditPath    string
	guardWrapper string
	auditWrapper string
}

func cloneAugmentObject(value map[string]any) map[string]any {
	clone := make(map[string]any, len(value))
	for key, item := range value {
		clone[key] = item
	}
	return clone
}

func validateAugmentHandler(
	value any,
	event string,
	groupIndex, handlerIndex int,
) (map[string]any, error) {
	label := fmt.Sprintf(
		"Auggie settings handler hooks.%s[%d].hooks[%d]",
		event, groupIndex, handlerIndex,
	)
	handler, ok := value.(map[string]any)
	if !ok || handler == nil {
		return nil, fmt.Errorf("%s must be an object", label)
	}
	handlerType, ok := handler["type"].(string)
	if !ok || handlerType != "command" {
		return nil, fmt.Errorf(`%s type must be "command"`, label)
	}
	command, ok := handler["command"].(string)
	if !ok || command == "" {
		return nil, fmt.Errorf("%s requires a non-empty command", label)
	}
	if arguments, exists := handler["args"]; exists {
		values, ok := arguments.([]any)
		if !ok {
			return nil, fmt.Errorf("%s args must be an array of strings", label)
		}
		for _, argument := range values {
			if _, ok := argument.(string); !ok {
				return nil, fmt.Errorf("%s args must be an array of strings", label)
			}
		}
	}
	if timeoutValue, exists := handler["timeout"]; exists {
		timeout, ok := timeoutValue.(float64)
		if !ok || math.IsNaN(timeout) || math.IsInf(timeout, 0) || timeout <= 0 {
			return nil, fmt.Errorf("%s timeout must be a positive finite number", label)
		}
	}
	return cloneAugmentObject(handler), nil
}

func validateAugmentMetadata(value any, label string) error {
	metadata, ok := value.(map[string]any)
	if !ok || metadata == nil {
		return fmt.Errorf("%s metadata must be an object", label)
	}
	for _, key := range []string{
		"includeConversationData",
		"includeMCPMetadata",
		"includeUserContext",
	} {
		if field, exists := metadata[key]; exists {
			if _, ok := field.(bool); !ok {
				return fmt.Errorf("%s metadata.%s must be a boolean", label, key)
			}
		}
	}
	return nil
}

func validateAugmentGroup(value any, event string, groupIndex int) (augmentGroup, error) {
	label := fmt.Sprintf("Auggie settings group hooks.%s[%d]", event, groupIndex)
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return augmentGroup{}, fmt.Errorf("%s must be an object", label)
	}
	if _, sessionEvent := augmentSessionEvents[event]; sessionEvent {
		if _, hasMatcher := object["matcher"]; hasMatcher {
			return augmentGroup{}, fmt.Errorf("%s matcher is only supported for tool events", label)
		}
	}
	if matcher, exists := object["matcher"]; exists {
		if _, ok := matcher.(string); !ok {
			return augmentGroup{}, fmt.Errorf("%s matcher must be a string", label)
		}
	}
	if metadata, exists := object["metadata"]; exists {
		if err := validateAugmentMetadata(metadata, label); err != nil {
			return augmentGroup{}, err
		}
	}
	values, ok := object["hooks"].([]any)
	if !ok {
		return augmentGroup{}, fmt.Errorf("%s must contain a hooks array", label)
	}
	handlers := make([]map[string]any, 0, len(values))
	for handlerIndex, value := range values {
		handler, err := validateAugmentHandler(value, event, groupIndex, handlerIndex)
		if err != nil {
			return augmentGroup{}, err
		}
		handlers = append(handlers, handler)
	}
	return augmentGroup{object: cloneAugmentObject(object), handlers: handlers}, nil
}

func readAugmentHooks(root map[string]any) (augmentHooks, error) {
	value, exists := root["hooks"]
	if !exists {
		return augmentHooks{}, nil
	}
	object, ok := value.(map[string]any)
	if !ok || object == nil {
		return nil, fmt.Errorf(`Auggie settings field "hooks" must be an object`)
	}
	hooks := make(augmentHooks, len(object))
	for event, groupValue := range object {
		if _, toolEvent := augmentToolEvents[event]; !toolEvent {
			if _, sessionEvent := augmentSessionEvents[event]; !sessionEvent {
				return nil, fmt.Errorf(
					`Auggie settings field "hooks.%s" uses an unsupported event`,
					event,
				)
			}
		}
		values, ok := groupValue.([]any)
		if !ok {
			return nil, fmt.Errorf(
				`Auggie settings field "hooks.%s" must be an array`,
				event,
			)
		}
		groups := make([]augmentGroup, 0, len(values))
		for groupIndex, value := range values {
			group, err := validateAugmentGroup(value, event, groupIndex)
			if err != nil {
				return nil, err
			}
			groups = append(groups, group)
		}
		hooks[event] = groups
	}
	return hooks, nil
}

func parseAugmentDocument(configPath string, raw []byte) (*augmentDocument, error) {
	label := fmt.Sprintf("Auggie user settings at %s", configPath)
	root, err := decodeStrictJSONObject(raw, label)
	if err != nil {
		return nil, err
	}
	hooks, err := readAugmentHooks(root)
	if err != nil {
		return nil, err
	}
	return &augmentDocument{
		exists: true, configPath: configPath, root: root, hooks: hooks,
		raw: append([]byte(nil), raw...),
	}, nil
}

func createAugmentDocument(configPath string) *augmentDocument {
	return &augmentDocument{
		configPath: configPath,
		root:       map[string]any{},
		hooks:      augmentHooks{},
	}
}

func renderAugmentDocument(
	document *augmentDocument,
	hooks augmentHooks,
) (*augmentRenderedDocument, error) {
	if document == nil {
		return nil, fmt.Errorf("Auggie settings document is required")
	}
	if reflect.DeepEqual(hooks, document.hooks) {
		return &augmentRenderedDocument{document: document}, nil
	}
	if !document.exists && len(hooks) == 0 {
		return &augmentRenderedDocument{document: document}, nil
	}
	root := cloneAugmentObject(document.root)
	if len(hooks) == 0 {
		delete(root, "hooks")
	} else {
		root["hooks"] = renderAugmentHooks(hooks)
	}
	if len(root) == 0 {
		return &augmentRenderedDocument{
			document: document,
			changed:  true,
			remove:   true,
		}, nil
	}
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode Auggie user settings: %w", err)
	}
	next = append(next, '\n')
	if _, err := parseAugmentDocument(document.configPath, next); err != nil {
		return nil, fmt.Errorf("validate rendered Auggie user settings: %w", err)
	}
	return &augmentRenderedDocument{
		document: document,
		changed:  !document.exists || !bytes.Equal(next, document.raw),
		next:     next,
	}, nil
}

func renderAugmentHooks(hooks augmentHooks) map[string]any {
	result := make(map[string]any, len(hooks))
	for event, groups := range hooks {
		values := make([]any, 0, len(groups))
		for _, group := range groups {
			object := cloneAugmentObject(group.object)
			handlers := make([]any, 0, len(group.handlers))
			for _, handler := range group.handlers {
				handlers = append(handlers, handler)
			}
			object["hooks"] = handlers
			values = append(values, object)
		}
		result[event] = values
	}
	return result
}

func readAugmentWindowsArgument(command string) (string, bool) {
	if len(command) < 2 || command[0] != '"' {
		return "", false
	}
	var value strings.Builder
	for index := 1; index < len(command); index++ {
		if command[index] == '\\' && index+1 < len(command) && command[index+1] == '"' {
			value.WriteByte('"')
			index++
			continue
		}
		if command[index] == '"' {
			return value.String(), index == len(command)-1 && value.Len() > 0
		}
		value.WriteByte(command[index])
	}
	return "", false
}

func readAugmentPOSIXArgument(command string) (string, bool) {
	if len(command) < 2 || command[0] != '\'' {
		return "", false
	}
	var value strings.Builder
	for index := 1; index < len(command); {
		if strings.HasPrefix(command[index:], augmentPOSIXApostrophe) {
			value.WriteByte('\'')
			index += len(augmentPOSIXApostrophe)
			continue
		}
		if command[index] == '\'' {
			return value.String(), index == len(command)-1 && value.Len() > 0
		}
		value.WriteByte(command[index])
		index++
	}
	return "", false
}

func parseAugmentCommand(command string) (string, bool) {
	if runtime.GOOS == "windows" {
		return readAugmentWindowsArgument(command)
	}
	return readAugmentPOSIXArgument(command)
}

func normalizeAugmentPath(value string) string {
	absolute, err := filepath.Abs(value)
	if err == nil {
		value = absolute
	}
	value = filepath.ToSlash(filepath.Clean(value))
	if runtime.GOOS == "windows" {
		return strings.ToLower(value)
	}
	return value
}

func sameAugmentPath(left, right string) bool {
	return normalizeAugmentPath(left) == normalizeAugmentPath(right)
}

func sameAugmentAgentID(left, right string) bool {
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func sameAugmentFileName(left, right string) bool {
	if runtime.GOOS == "windows" {
		return strings.EqualFold(left, right)
	}
	return left == right
}

func managedAugmentAgentID(
	handler map[string]any,
	wrapperName, runtimeRoot string,
) (string, bool) {
	if handler["type"] != "command" || handler["timeout"] != augmentHookTimeout {
		return "", false
	}
	if _, hasArguments := handler["args"]; hasArguments {
		return "", false
	}
	command, ok := handler["command"].(string)
	if !ok {
		return "", false
	}
	wrapperPath, ok := parseAugmentCommand(command)
	if !ok || !sameAugmentFileName(filepath.Base(wrapperPath), wrapperName) {
		return "", false
	}
	agentDirectory := filepath.Dir(wrapperPath)
	if !sameAugmentPath(filepath.Dir(agentDirectory), runtimeRoot) {
		return "", false
	}
	agentID := filepath.Base(agentDirectory)
	return agentID, agentID != "" && agentID != "." && agentID != ".."
}

func removeManagedAugmentGroups(
	groups []augmentGroup,
	wrapperName, agentID, runtimeRoot string,
) ([]augmentGroup, bool) {
	result := make([]augmentGroup, 0, len(groups))
	changed := false
	for _, group := range groups {
		handlers := make([]map[string]any, 0, len(group.handlers))
		groupChanged := false
		for _, handler := range group.handlers {
			managedID, managed := managedAugmentAgentID(handler, wrapperName, runtimeRoot)
			remove := managed && (agentID == "" || sameAugmentAgentID(managedID, agentID))
			if remove {
				changed = true
				groupChanged = true
				continue
			}
			handlers = append(handlers, handler)
		}
		if len(handlers) > 0 || !groupChanged {
			result = append(result, augmentGroup{object: group.object, handlers: handlers})
		}
	}
	return result, changed
}

func removeManagedAugmentHooks(
	hooks augmentHooks,
	agentID, runtimeRoot string,
) (augmentHooks, bool) {
	result := make(augmentHooks, len(hooks))
	for event, groups := range hooks {
		result[event] = groups
	}
	changed := false
	for _, contract := range []struct{ event, wrapper string }{
		{"PreToolUse", augmentGuardWrapperName()},
		{"PostToolUse", augmentAuditWrapperName()},
	} {
		groups, eventChanged := removeManagedAugmentGroups(
			result[contract.event], contract.wrapper, agentID, runtimeRoot,
		)
		if !eventChanged {
			continue
		}
		changed = true
		if len(groups) == 0 {
			delete(result, contract.event)
		} else {
			result[contract.event] = groups
		}
	}
	return result, changed
}
